pgdedupe/mssql/cli.py

```diff
@@ -0,0 +1,526 @@
+        with open(config['training_file']) as tf:
+            deduper.readTraining(tf)
+
+    if config['prompt_for_labels']:
+        # ## Active learning
+        print('starting active labeling...')
+        # Starts the training loop. Dedupe will find the next pair of records
+        # it is least certain about and ask you to label them as duplicates
+        # or not.
+
+        # use 'y', 'n' and 'u' keys to flag duplicates
+        # press 'f' when you are finished
+        dedupe.convenience.consoleLabel(deduper)
+        # When finished, save our labeled, training pairs to disk
+        with open(config['training_file'], 'w') as tf:
+            deduper.writeTraining(tf)
+
+    # `recall` is the proportion of true dupes pairs that the learned
+    # rules must cover. You may want to reduce this if your are making
+    # too many blocks and too many comparisons.
+    deduper.train(recall=config['recall'])
+
+    with open(config['settings_file'], 'wb') as sf:
+        deduper.writeSettings(sf)
+
+    # We can now remove some of the memory hobbing objects we used
+    # for training
+    deduper.cleanupTraining()
+    return deduper
+
+
+# Blocking
+def create_blocking(deduper, con, config):
+    c = con.cursor()
+
+    # To run blocking on such a large set of data, we create a separate table
+    # that contains blocking keys and record ids
+    print('creating blocking_map database')
+    c.execute("IF OBJECT_ID('{schema}.blocking_map', 'U') IS NOT NULL "
+              "DROP TABLE {schema}.blocking_map".format(**config))
+    c.execute("CREATE TABLE {schema}.blocking_map "
+              "(block_key VARCHAR(200), _unique_id INT)".format(**config))  # TODO: THIS INT...
+    # ... needs to be dependent upon the column type of entry_id
+
+    # If dedupe learned a Index Predicate, we have to take a pass
+    # through the data and create indices.
+    print('creating inverted index')
+
+    for field in deduper.blocker.index_fields:
+        c2 = con.cursor(as_dict=True)
+        c2.execute("SELECT DISTINCT {0} FROM {schema}.entries_unique".format(field, **config))
+        field_data = (unicode_to_str(row)[field] for row in c2)
+        deduper.blocker.index(field_data, field)
+        c2.close()
+
+    # Now we are ready to write our blocking map table by creating a
+    # generator that yields unique `(block_key, donor_id)` tuples.
+    print('writing blocking map')
+
+    c3 = con.cursor(as_dict=True)
+    c3.execute("SELECT {all_columns} FROM {schema}.entries_unique".format(**config))
+    full_data = ((row['_unique_id'], unicode_to_str(row)) for row in c3)
+    b_data = deduper.blocker(full_data)
+
+    # Write out blocking map to CSV so we can quickly load in with
+    csv_file = tempfile.NamedTemporaryFile(prefix='blocks_', delete=False, mode='w')
+    csv_writer = csv.writer(csv_file)
+    csv_writer.writerows(b_data)
+    c3.close()
+    csv_file.close()
+
+    # write from csv to SQL
```

Comment:
I've not looked — does MSSQL not have a copy-from-CSV intrinsic?  If not, it'd probably be simpler and quicker to just skip the CSV entirely.  The only reason we do this for the postgres is because the batch copy from CSV is orders of magnitude faster than incremental row-by-row updates.